zerver/views/auth.py

```diff
@@ -330,10 +330,30 @@ def remote_user_sso(request: HttpRequest,
     if not auth_enabled_helper([ZulipRemoteUserBackend.auth_backend_name], realm):
         return redirect_to_config_error("remoteuser/backend_disabled")
 
-    try:
-        remote_user = request.META["REMOTE_USER"]
-    except KeyError:
-        return redirect_to_config_error("remoteuser/remote_user_header_missing")
+    if not settings.SSO_HTTP_USER_HEADER:
+        # Look for REMOTE_USER environment variable header
+        try:
+            remote_user = request.META["REMOTE_USER"]
+        except KeyError:
+            return redirect_to_config_error("remoteuser/remote_user_header_missing")
+    else:
+        if settings.SSO_HTTP_SECRET_HEADER:
+            # Check the shared secret header exists and contains the correct value
+            try:
+                if request.headers[settings.SSO_HTTP_SECRET_HEADER] != settings.SSO_HTTP_SECRET:
+                    logging.warning(
+                        f"SSO shared secret doesn't match value from {settings.SSO_HTTP_SECRET_HEADER}")
+                    return redirect_to_config_error("remoteuser/remote_user_http_secret_incorrect")
+            except KeyError:
+                logging.warning(
+                    f"SSO request missing shared secret header: {settings.SSO_HTTP_SECRET_HEADER}")
+                return redirect_to_config_error("remoteuser/remote_user_http_secret_missing")
+
+        logging.debug(f"Looking for SSO HEADER {settings.SSO_HTTP_USER_HEADER}")
+        try:
+            remote_user = request.headers[settings.SSO_HTTP_USER_HEADER]
+        except KeyError:
+            return redirect_to_config_error("remoteuser/remote_user_http_header_missing")
```

Comment:
Can we make the code paths a bit less duplicated, by just using `REMOTE_USER` as the default header name?  I think that'd make this a lot simpler.

(Structurally, one way to do that is have `default_settings.py` set `SSO_HTTP_USER_HEADER="REMOTE_USER"`.

Reply:
`REMOTE_USER` is not a header.

Reply:
That was my initial approach, but the headers are treated differently when sent from a remote host. E.g. if you send an HTTP header `REMOTE_USER` from a remote host the [request.META](https://docs.djangoproject.com/en/3.0/ref/request-response/#django.http.HttpRequest.META) field will be prepended with `HTTP_` and transformed to uppercase with `-` replaced with `_`, so you get `HTTP_REMOTE_USER`.

I originally looked for `SSO_HTTP_SECRET_HEADER` in `request.META` after transforming it `.upper().replace('-', '_')`, but thought having separate paths was cleaner for a few reasons. If we were to use `request.META` users wanting to use HTTP headers would be required to prepend them with `HTTP_`, otherwise if we did that ourselves we could silently open up existing users of the environment variable to masquerading attacks from remote hosts.

E.g. Sending the successful curl login:
request.META:
```json
{
<SNIP>
    "HTTP_ACCEPT": "*/*",
    "HTTP_CONNECTION": "close",
    "HTTP_HOST": "localhost:9991",
    "HTTP_USER_AGENT": "curl/7.54.0",
    "HTTP_X_FORWARDED_PORT": "9991",
    "HTTP_X_REAL_IP": "172.17.0.1",
    "HTTP_X_REMOTE_SECRET": "secretval",
    "HTTP_X_REMOTE_USER": "iago@zulip.com",
<SNIP>
}
```

requests.headers:
```json
{
<SNIP>
    "Accept": "*/*",
    "Connection": "close",
    "Content-Length": "0",
    "Content-Type": "text/plain",
    "Host": "localhost:9991",
    "User-Agent": "curl/7.54.0",
    "X-Forwarded-Port": "9991",
    "X-Real-Ip": "172.17.0.1",
    "X-Remote-Secret": "secretval",
    "X-Remote-User": "iago@zulip.com",
<SNIP>
}
```